### PR TITLE
[#4] Content-assist for Swagger

### PR DESCRIPTION
**Pull request by tfesenko**
The main part is implemented. There are several remaining minor issues described in https://github.com/ModelSolv/SwagEdit/issues/4#issuecomment-156177871:
- content assist inside an object that only defines pattern properties.
- distinguishing required properties in the content assist popup.
- better looking labels in the popup (e.g. use of different colors)
## Implementation

screen shot 2015-11-12 at 4 20 10 pm
## Tests
- `SwaggerProposalProviderTest`
- `SwaggerSchemaTest`

### CODE REVIEW

File: com.reprezen.swagedit/src/com/reprezen/swagedit/assist/JsonUtil.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+			return "enum";
+		}
+		else if (node.has("type")) {
+			return node.get("type").asText();
+		} 
+		else if (isRef(node)) {
+			return getType(document, getRef(document, node));
+		} else if (node.has("properties")) {
+			return "object";
+		}
```
Comment (tfesenko) [inline com.reprezen.swagedit/src/com/reprezen/swagedit/assist/JsonUtil.java]: Code formatting is inconsistent here. I suggest to choose a formatter and always use it.

File: com.reprezen.swagedit/src/com/reprezen/swagedit/assist/SwaggerProposalProvider.java [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+import com.fasterxml.jackson.databind.node.ArrayNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.google.common.base.Predicate;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Sets;
+
+/**
+ * 
+ * 
+ *
```
Comment (tfesenko) [inline com.reprezen.swagedit/src/com/reprezen/swagedit/assist/SwaggerProposalProvider.java]: An empty javadoc is fine for now. Just let's make sure that we add an actual description or remove it later.

File: com.reprezen.swagedit/src/com/reprezen/swagedit/assist/SwaggerProposalProvider.java [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+	 * @param offset
+	 * @return list of completion proposals
+	 */
+	public List<ICompletionProposal> getProposals(JsonNode proposal, String prefix, int offset) {
+		if (proposal == null || !proposal.has(TYPE))
+			return Collections.emptyList();
+
+		if (prefix != null && prefix.trim().isEmpty()) {
+			prefix = null;
+		}
```
Comment (tfesenko) [inline com.reprezen.swagedit/src/com/reprezen/swagedit/assist/SwaggerProposalProvider.java]: If the prefix can contain leading or trailing traces we have to trim them. What about replacing it with this then:

``` java
prefix = prefix.trim();
Strings.nullToEmpty(prefix);
```

File: com.reprezen.swagedit/src/com/reprezen/swagedit/assist/SwaggerProposalProvider.java [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+
+		if (type != null) {
+			switch (type) {
+			case "object":
+				return createObjectProposal(schema, data, definition);
+			case "string":
+				return createStringProposal(schema, data, definition);
+			case "enum":
+				return createEnumProposal(schema, data, definition);
+			case "oneOf":
```
Comment (tfesenko) [inline com.reprezen.swagedit/src/com/reprezen/swagedit/assist/SwaggerProposalProvider.java]: We use `"oneOf"` five times in this class and once in JsonUtils. Why don't we extract it (and other similar JSON Schema strings) as a constant?

File: com.reprezen.swagedit/src/com/reprezen/swagedit/validation/SwaggerSchema.java [hunk 5]
```diff
@@ -47,45 +46,117 @@ public JsonSchema getSchema() {
...
-			rootKeywords.addAll(collectKeywords(getTree()));
-			keywords.addAll(rootKeywords);
+	/**
+	 * Returns the proposal that matches the given path in the document.
+	 * 
+	 * @param path
+	 * @param document
+	 * @return proposal
+	 */
+	public JsonNode getProposals(String path, JsonNode document) {
```
Comment (tfesenko) [inline com.reprezen.swagedit/src/com/reprezen/swagedit/validation/SwaggerSchema.java]: I am not sure this name accurately reflects what this method does. It returns a single node, but the method name uses a plural form. What about `getProposalNode`?

File: com.reprezen.swagedit/src/com/reprezen/swagedit/assist/SwaggerProposalProvider.java [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+
+	/**
+	 * Returns proposal for the given data and schema definition.
+	 * 
+	 * @param schema
+	 * @param data
+	 * @param definition
+	 * @return proposal
+	 */
+	public JsonNode get(JsonNode schema, JsonNode data, JsonNode definition) {
```
Comment (tfesenko) [inline com.reprezen.swagedit/src/com/reprezen/swagedit/assist/SwaggerProposalProvider.java]: What about "`getProposalContextNode`" as a method name?

Comment (tfesenko) [inline com.reprezen.swagedit/src/com/reprezen/swagedit/assist/SwaggerProposalProvider.java]: Also, this method is only called from `SwaggerSchema` (and tests). What about moving it along with the create*Proposal() methods and the `mapper` to `SwaggerSchema`?

File: com.reprezen.swagedit/src/com/reprezen/swagedit/validation/SwaggerSchema.java [hunk 7]
```diff
@@ -33,7 +32,7 @@
 	 */
 	public JsonSchema getSchema() {
 		if (schema == null) {
```
Comment (tfesenko) [inline com.reprezen.swagedit/src/com/reprezen/swagedit/validation/SwaggerSchema.java:30]: Nice caching.

**Comment by tfesenko** [comment]
Looks great!
@ghillairet , I left several comments, can you please address them to make it even better?

File: com.reprezen.swagedit/src/com/reprezen/swagedit/assist/SwaggerProposalProvider.java [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+				return createArrayProposal(schema, data, definition);
+			default:
+				return null;
+			}
+		}
+
+		return null;
+	}
+
+	private JsonNode createArrayProposal(JsonNode schema, JsonNode data, JsonNode definition) {
```
Comment (tfesenko) [inline com.reprezen.swagedit/src/com/reprezen/swagedit/assist/SwaggerProposalProvider.java]: The method name is confusing, it makes me think that it returns a JFace `ICompletionProposal`, but it's a JSON node. What about `createArrayProposalNode()`? Similar with other `create` methods.

File: com.reprezen.swagedit.tests/src/com/reprezen/swagedit/tests/SwaggerProposalProviderTest.xtend [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+	}
+
+	@Test
+	def void testGetProposalForEnumDefinition() {
+		val definition = schema.asJson.get("properties").get("swagger")		
+		val proposal = provider.get(schema.asJson, mapper.createObjectNode, definition)
+
+		assertNotNull(proposal)
+		assertEquals("enum", proposal.get(SwaggerProposalProvider.TYPE).asText)
+		assertArrayEquals(#["2.0"], proposal.get("literals").map(it | it.asText))
```
Comment (tfesenko) [inline com.reprezen.swagedit.tests/src/com/reprezen/swagedit/tests/SwaggerProposalProviderTest.xtend]: Nice

File: com.reprezen.swagedit.tests/src/com/reprezen/swagedit/tests/SwaggerDocumentTest.xtend [hunk 10]
```diff
@@ -2,71 +2,117 @@ package com.reprezen.swagedit.tests
...
-		val path = document.getPath(1)
-		println(path)
-
-		println(document.getPath(3))
-		val node = Yaml.mapper.readTree(yaml)
-		println(node)
-		//val events = document.getEvent(1)		
-		//println(events)
+		
+		assertEquals(":paths:/", document.getPath(6, 3))
```
Comment (tfesenko) [inline com.reprezen.swagedit.tests/src/com/reprezen/swagedit/tests/SwaggerDocumentTest.xtend:115]: Nice

File: com.reprezen.swagedit/src/com/reprezen/swagedit/assist/JsonUtil.java (hunk 1 above)
Comment (tedepstein) [inline com.reprezen.swagedit/src/com/reprezen/swagedit/assist/JsonUtil.java]: For easy coordination across the team, shouldn't we use our standard ModSquad formatter config?

File: com.reprezen.swagedit/src/com/reprezen/swagedit/validation/SwaggerSchema.java (hunk 5 above)
Comment (ghillairet) [inline com.reprezen.swagedit/src/com/reprezen/swagedit/validation/SwaggerSchema.java]: I have removed that method. 
SwaggerProposalProvider is now the only class that should be use to obtain completion proposals.

File: com.reprezen.swagedit/src/com/reprezen/swagedit/assist/SwaggerProposalProvider.java (hunk 6 above)
Comment (ghillairet) [inline com.reprezen.swagedit/src/com/reprezen/swagedit/assist/SwaggerProposalProvider.java]: I renamed it createProposals as it now returns a list of JsonNodes.
The previous createProposal is now renamed getCompletionProposals() and is the only one that need to be called by SwaggerContentAssistProcessor to obtain the list of ICompletionProposal.

File: com.reprezen.swagedit/src/com/reprezen/swagedit/assist/SwaggerProposalProvider.java (hunk 3 above)
Comment (ghillairet) [inline com.reprezen.swagedit/src/com/reprezen/swagedit/assist/SwaggerProposalProvider.java]: Strings.nullToEmpty(prefix) is being used now.

File: com.reprezen.swagedit/src/com/reprezen/swagedit/assist/SwaggerProposalProvider.java (hunk 4 above)
Comment (ghillairet) [inline com.reprezen.swagedit/src/com/reprezen/swagedit/assist/SwaggerProposalProvider.java]: There is now an enum JsonType that contains all those.

**Comment by tfesenko** [comment]
QA: I like the content assist in general, there are just several improvements I suggest so far:
1. Content-assist for boolean properties does not work
In the model created by the wizard navigate to the line 29: `required: true`.  Remove the value and navigate cursor after the colon: `required: //cursor`. Nothing will be shown
2. String values are wrapped in single quotes `'value'` while the model template uses double quotes - `"value"`

**Comment by tfesenko** [comment]
I like the new code. It's concise and readable. Merging.
@ghillairet , I left several comments, can you please address them?

**Pull request merged**